File: app/src/main/java/com/adsamcik/signalcollector/utility/Parser.java
package com.adsamcik.signalcollector.utility;

import android.content.Context;
import android.support.annotation.NonNull;

import com.adsamcik.signalcollector.file.DataStore;
import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class Parser {

	/**
	 * Tries to parse json to object
	 *
	 * @param json   json
	 * @param tClass object class
	 * @param <T>    object type
	 * @return object if success, nul otherwise
	 */
	public static <T> T tryFromJson(String json, Class<T> tClass) {
		if (json != null && !json.isEmpty()) {
			try {
				return new Gson().fromJson(json, tClass);
			} catch (JsonSyntaxException e) {
				FirebaseCrash.report(e);
			}
		}
		return null;
	}


	public static ArrayList<String[]> parseTSVFromFile(@NonNull Context context, @NonNull String fileName) {
		if (DataStore.exists(context, fileName)) {
			ArrayList<String[]> items = new ArrayList<>();
			try (FileInputStream fis = context.openFileInput(fileName)) {
				InputStreamReader isr = new InputStreamReader(fis);
				BufferedReader br = new BufferedReader(isr);
				String receiveString;

				while ((receiveString = br.readLine()) != null)
					items.add(parseLine(receiveString));

				isr.close();
				return items;
			} catch (IOException e) {
				FirebaseCrash.report(e);
			}
		}

		return null;
	}

	private static String[] parseLine(String line) {
		if (line == null || line.length() == 0)
			return null;
		return line.split("\t");
	}


}

File: app/src/main/java/com/adsamcik/signalcollector/utility/Parser.kt
package com.adsamcik.signalcollector.utility

import android.util.Log
import com.adsamcik.signalcollector.BuildConfig
import com.crashlytics.android.Crashlytics
import com.squareup.moshi.JsonDataException
import com.squareup.moshi.Moshi
import java.io.IOException



/**
 * Utility tool for parsing different types of files to objects
 */
object Parser {

    /**
     * Tries to parse json to object
     *
     * @param json   json
     * @param tClass object class
     * @param <T>    object type
     * @return object if successful, null otherwise
    </T> */
    fun <T> tryFromJson(json: String?, tClass: Class<T>): T? {
        if (json != null && !json.isEmpty()) {
            val moshi = Moshi.Builder().build()
            try {
                return moshi.adapter(tClass).fromJson(json)
            } catch (e: IOException) {
                if (BuildConfig.DEBUG)
                    Log.e("FATAL EXCEPTION", json)
                else
                    Crashlytics.logException(e)
            } catch (e: JsonDataException) {
                if (BuildConfig.DEBUG)
                    Log.e("FATAL EXCEPTION", "type ${tClass.name} data $json")
                else
                    Crashlytics.logException(e)
            }

        }
        return null
    }

    private fun parseLine(line: String): Array<String>? =
            if (line.isEmpty()) null else line.split("\t".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()


}
